### PR TITLE
GH-109369: Exit tier 2 if executor is invalid

### LINKED ISSUE #109369: Executors might ignore instrumentation.
# Bug report

Code handed to the optimizer may not include instrumentation. If instrumentation is added later, the executor does not see it.
We remove all `ENTER_EXECUTORS` when instrumenting, but that doesn't fix the problem of executors that are still running.

### Example

A loop that calls `foo`:
```Py
while large_number > 0:
    foo()
    large_number -= 1
```
The loop gets turned into an executor, and sometime before `large_number` reaches zero, a debugger gets attached and in some callee of `foo` turns on monitoring of calls. We would expect all subsequent calls to `foo()` to be monitored, but they will not be, as the executor knows nothing about the call to `foo()` being instrumented.

### Let's not rely on the executor/optimizer to handle this.

We could add a complex de-optimization strategy, so that executors are invalidated when instrumentation occurs, but that is the sort of thing we want to be doing for maximum performance, not for correctness.

It is much safer, and ultimately no slower (once we implement fancy de-optimization) to add extra checks for instrumentation, so that unoptimized traces are correct.

### The solution

The solution is quite simple, add a check for instrumentation after every call.

We can make this less slow (and less simple) by combining the eval-breaker check and instrumentation check into one. This makes the check slightly more complex, but should speed up `RESUME` by less than it slows down every call as every call already has an eval-breaker check.

Combining the two checks reducing the number of bits for versions from 64 to 24 (we need the other bits for eval-breaker, gc, async exceptions, etc). A 64 bit number never overflows (computers don't last long enough), but 24 bit numbers do.

This will have three main effects, beyond the hopefully very small performance impact for all code:
* It removes the need for a complete call stack scan of all threads when instrumenting. 
* Tools that set or change monitoring many times will see significant performance improvements as we no longer need to traverse all stacks to re-instrument the whole call stack whenever monitoring changes
* Tools that set or change monitoring many millions of times will break, as we run out of versions. It is likely that these tools were already broken or had performance so bad as to be totally unusable.

#### Making the check explicit in the bytecode.

We can simplify all the call instructions by removing the `CHECK_EVAL_BREAKER` check at the end and adding an explicit `RESUME` instruction after every `CALL`

Although this has the disadvantage of making the bytecode larger and adding dispatch overhead, it does have the following advantages:
* Allows tier 1 to optimize the `RESUME` to `RESUME_CHECK` which might be cancel out the additional dispatch overhead.
* Makes the check explicit, making it feasible for tier 2 optimizations to remove it.

### Linked PRs
* gh-109846
* gh-110358
* gh-110384
* gh-111642
* gh-111657

### PR DESCRIPTION
**Pull request by markshannon**
Fixes GH-109369.
Adds a `_CHECK_VALIDITY` uop which is inserted after any escaping operations.
Ensures that uop traces will be exited promptly if the code is instrumented.

### CODE REVIEW

File: Python/optimizer.c [hunk 1]
```diff
@@ -464,8 +464,8 @@ translate_bytecode_to_trace(
     } \
     reserved = (n);  // Keep ADD_TO_TRACE / ADD_TO_STUB honest
 
-// Reserve space for main+stub uops, plus 2 for _SET_IP and _EXIT_TRACE
-#define RESERVE(main, stub) RESERVE_RAW((main) + (stub) + 2, uop_name(opcode))
+// Reserve space for main+stub uops, plus 2 for _SET_IP, _CHECK_VALIDITY and _EXIT_TRACE
```
Comment (iritkatriel) [inline Python/optimizer.c]:
```suggestion
// Reserve space for main+stub uops, plus 3 for _SET_IP, _CHECK_VALIDITY and _EXIT_TRACE
```

File: Tools/cases_generator/flags.py [hunk 2]
```diff
@@ -50,6 +100,10 @@ def fromInstruction(instr: parsing.Node) -> "InstructionFlags":
                 or variable_used(instr, "exception_unwind")
                 or variable_used(instr, "resume_with_error")
             ),
+            HAS_ESCAPES_FLAG=(
+                variable_used(instr, "tstate")
+                or makes_escaping_api_call(instr)
+            ),
```
Comment (iritkatriel) [inline Tools/cases_generator/flags.py:106]: The InstructionFlags were intended to be for things that get emitted into the metadata and used by the interpreter.  There is a bitmask created for them below. I think if we want to have things here which are only used in the generator then we need to make sure we don't waste a bit in that bit mask for them.

Comment (markshannon) [reply Tools/cases_generator/flags.py:106]: It gets used by the optimizer, so it needs to go in a table somewhere. This seems to be the only place to put it.
We ought to have separate tables for the tier 1 instructions, and for uops. But that's for another PR.

File: Python/optimizer_analysis.c [hunk 3]
```diff
@@ -36,6 +46,9 @@ remove_unneeded_uops(_PyUOpInstruction *buffer, int buffer_size)
             if (_PyOpcode_opcode_metadata[opcode].flags & (HAS_ERROR_FLAG | HAS_DEOPT_FLAG) || opcode == _PUSH_FRAME) {
                 need_ip = true;
             }
+            if (_PyOpcode_opcode_metadata[opcode].flags & HAS_ESCAPES_FLAG) {
```
Comment (iritkatriel) [inline Python/optimizer_analysis.c:49]: use OPCODE_HAS_ESCAPES macro

**Review by iritkatriel — APPROVED** [review]

**Comment by brandtbucher** [comment]
So, on the JIT branch, we don't actually keep the executor handy when executing instructions (only when entering/exiting traces). I'm guessing we'll need it a lot more in the future once we start doing stitching and such, but I wonder: can we just stick it on the `tstate` or the `frame`? It feels a bit painful to dedicate a register to this thing, especially since it'll be hardly used in optimized traces (I think).

Not a big deal though, and can be deferred to a later date. Just something to ponder, or to do if it's easy now.

**Comment by brandtbucher** [comment]
Idea: put the executor's address in `_CHECK_VALIDITY`'s `operand`?

**Comment by brandtbucher** [comment]
Ah, never mind, the JIT can burn in the executor. I withdraw my comments. :)

**Pull request merged**